### PR TITLE
Add property row for single-row tables

### LINKED ISSUE #93: Need a better way to extract the contents of one-row tables
After `t.sum()` or some other reduction to a single row, it's messy to access the contents of the result using `t.rows[0]...`.

### PR DESCRIPTION
**Pull request by mDibyo**
This resolves #93.

### CODE REVIEW

File: datascience/tables.py [hunk 1]
```diff
@@ -202,6 +202,11 @@ def rows(self):
         return self.Rows(self)
 
     @property
+    def row(self):
+        assert self.num_rows == 1, 'Table has more than one row.'
+        return self.rows[0]
```
Comment (SamLau95) [inline datascience/tables.py]: Let's write a docstring for this method! And then add it to `docs/tables.rst`.

Comment (mDibyo) [inline datascience/tables.py]: Done!
`ValueError` does not seem the right choice, as its used to denote an error in input args. Switched to a `RuntimeError` instead.

**Comment by stefanv** [comment]
This is a very specialized method to add, and only applicable to a small subset of tables.  I am working on another "proof of concept" PR that would add ``t.rows[0]`` as a syntax for this.  I'll make a PR later today, then we can discuss it over there, if that's of any interest.

**Comment by mDibyo** [comment]
@stefanv, I am not sure I am following you. Are you working on a PR for an alternate way to get `table.rows[0]`?

File: tests/test_tables.py [hunk 2]
```diff
@@ -86,6 +86,23 @@ def test_basic_rows(t):
...
+
+def test_row():
+    t = Table([[1], [2]], ['one', 'two'])
+    assert_equal(t.row, 'Row(one=1, two=2)')
+
+
+def test_row_when_multiple_rows():
+    t = Table([[1, 3], [2, 4]], ['one', 'two'])
+    with pytest.raises(RuntimeError):
+        _ = t.row
```
Comment (SamLau95) [inline tests/test_tables.py]: Thanks for writing these!

**Comment by SamLau95** [comment]
Waiting on @papajohn and @stefanv to respond since I'm neutral about adding this method to the API.

**Comment by stefanv** [comment]
Yes, my feeling (without knowing the background of the Row class) is that it is not really necessary.  I will soon file an exploratory PR that refactors ``rows`` to return a table, and then ``rows[0]`` with give you the single row required.  I don't think it's worth adding a new method *just* to return the first row.

Sorry, folks, if I am introducing noise into these conversations—I know I am not up to date with the discussions had so far, but I hope to catch up.

**Comment by mDibyo** [comment]
@stefanv That's no problem! :)
Well `table.rows[0]` does indeed work right now. We are introducing `table.row` specifically for single-row table to make it easier to get the single row. Check out discussion at #93.
To get rows as a separate table, we already have `Table.take`. Do you think we should not have a separate method for getting just individual data points (not in table format)?

File: tests/test_tables.py (hunk 2 above)
Comment (mDibyo) [inline tests/test_tables.py]: Ohhh, don't thank me. I should always be writing tests for any method I write...

**Comment by deculler** [comment]
John had pretty strong feelings about the row design, so I'll let him weigh
in here.  Seems like you are working with the Tables_as_matrix concept
rather than ordered collection of named columns.  A row is really like a
record, or rather Dict.  But of course it can be a sequence too, since the
columns are ordered.  This also preserves the connection with relations -
tables in SQL.  So I would say the design is not a random accident.  It was
pretty carefully developed as a concept.  The teaching goal helps keep that
in focus.  Yes there are lots of implementations, and the row
implementation has gone through some changes.

**Comment by stefanv** [comment]
The way I think about it, a Table is a container of columns.  If you then extract entries from the table, you should again get a multi-column object (which, essentially, is a Table again).  If you think of a Table as a container of rows, the abstraction described above fits more easily.  But I think trying to accommodate both can be quite confusing.  So, what I'm implementing is a fairly minor change that simply removes the notion of a Row, but still gives you a Table object when selecting a row—and that object still has access to columns, as expected.

**Comment by stefanv** [comment]
After some more experimenting, the above seems to be doable using array scalars, but it does break down on the notion of ``table.rows[0]`` giving a single object and ``table.rows[:2]`` returning a container (as you would get in, e.g., numpy and other N-D containers).  So perhaps then the smallest change that would make sense is to simply alter ``table.rows`` to accept a slice and then return a Table instead of multiple rows.

**Comment by stefanv** [comment]
That approach turned out to be a *lot* easier: https://github.com/dsten/datascience/pull/119

Which makes me think it's perhaps closer to the right pattern.

**Comment by stefanv** [comment]
See also #120 for a different idea (adding slicing to `take`)

**Comment by deculler** [comment]
I definitely like the idea of adding slicing to take.  It is frustrating to
have to write the enumerator there.

David E. Culler
Friesen Professor of Computer Science
Electrical Engineering and Computer Sciences
University of California, Berkeley

**Comment by SamLau95** [comment]
Pinging @papajohn ; I agree that this method is very specialized and it's always much harder to remove a method than to add one. Do you see this being useful enough to add it to the API?

**Comment by papajohn** [comment]
Let's drop the single-row tables special support.

**Pull request closed without merging**